src/DomainGuest.sol

```diff
@@ -0,0 +1,365 @@
+    function _int256(uint256 x) internal pure returns (int256 y) {
+        require((y = int256(x)) >= 0, ARITHMETIC_ERROR);
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+    function _max(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x >= y ? x : y;
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        unchecked {
+            z = x != 0 ? ((x - 1) / y) + 1 : 0;
+        }
+    }
+
+    // --- Administration ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "end") end = EndLike(data);
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "dust") dust = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Mint DAI and send to user
+    /// @param to The address to send the DAI to on the local domain
+    /// @param amount The amount of DAI to send [WAD]
+    function _deposit(address to, uint256 amount) internal {
+        vat.swell(address(this), _int256(amount * RAY));
+        daiJoin.exit(to, amount);
+
+        emit Deposit(to, amount);
+    }
+
+    /// @notice Withdraw DAI by burning local canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to withdraw [WAD]
+    function _withdraw(address to, uint256 amount) internal {
+        require(dai.transferFrom(msg.sender, address(this), amount), "DomainGuest/transfer-failed");
+        daiJoin.join(address(this), amount);
+        vat.swell(address(this), -_int256(amount * RAY));
+
+        emit Withdraw(msg.sender, to, amount);
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Record changes in line and grain and update dss global debt ceiling if necessary
+    /// @param _lid Local ordering id
+    /// @param wad The new debt ceiling [WAD]
+    function _lift(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 lastLine = vat.Line();
+        uint256 rad = wad * RAY;
+        if (rad > lastLine) grain = wad;
+        vat.file("Line", rad);
+
+        emit Lift(wad);
+    }
+
+    /// @notice Will release remote DAI from the escrow when it is safe to do so
+    /// @dev    Should be run by keeper on a regular schedule.
+    function _release() internal returns (uint256 _rid, uint256 _burned) {
+        require(live == 1, "DomainGuest/not-live");
+
+        uint256 limit = _max(vat.Line() / RAY, _divup(vat.debt(), RAY));
```

Comment:
Leaving a comment here which is probably part of a bigger discussion.
`limit` should also consider a buffer over the debt. We do not want (at least in normal conditions) leave the `grain` right to the limit to the `debt` as the second after we will have unbacked DAI.
As a I said, we need to think which kind of mechanism we should put in place to avoid this ever happens.

Reply:
I think including a buffer makes sense.

Reply:
Ok good, will think about it and try to get a PR with a proposed solution.